Importing data

In [1]:
import pandas

In [6]:
demo_covid = pandas.read_csv('demo_covid.csv')

Splitting data into training & Test

In [7]:
demo_covid_train = demo_covid.loc[:499,:]

In [8]:
demo_covid_test = demo_covid.loc[500:,:]

Alos sorting the train data classwise

In [9]:
pos_covid = demo_covid_train[demo_covid_train['9']=='pos']

In [10]:
neg_covid = demo_covid_train[demo_covid_train['9']=='neg']

Train the model{Binary classifier,normal,(Bays theory)multivariate,(Multivariate parameter),qudratic discriminator by putting decision threshold 0.5 on posterior probability}.

Bays theory, post_pos= Likelihood_pos * Prior_Pos/Likelihood_pos* Prior_pos + Likelihood_neg* Prior_neg

Multivariate, f(x)= 1/(2pi**p/2 * det sig**1/2) e**-1/2(x-mu)**T * sig**-1(x-mu); we need two parameter mean=mu and cov=sigma

In [11]:
from scipy.stats import multivariate_normal

In [12]:
mu_pos = pos_covid.mean()
mu_neg = neg_covid.mean()

In [13]:
sig_pos = pos_covid.cov()
sig_neg = neg_covid.cov()

In [15]:
multivariate_normal_pos = multivariate_normal(mu_pos,sig_pos)
multivariate_normal_neg = multivariate_normal(mu_neg,sig_neg)

In [17]:
demo_covid_train_withoutlebel= demo_covid_train.drop(['9'],axis=1)

In [18]:
prior_pos=len(pos_covid)/(len(pos_covid)+len(neg_covid))
prior_neg=len(neg_covid)/(len(pos_covid)+len(neg_covid))

In [21]:
#post_pos = multivariate_normal_pos.pdf(demo_covid_train_withoutlebel)*prior_pos/multivariate_normal_pos.pdf(demo_covid_train_withoutlebel)*prior_pos+multivariate_normal_neg.pdf(demo_covid_train_withoutlebel)*prior_neg
#post_neg = multivariate_normal_neg.pdf(demo_covid_train_withoutlebel)*prior_neg/multivariate_normal_neg.pdf(demo_covid_train_withoutlebel)*prior_neg+multivariate_normal_pos.pdf(demo_covid_train_withoutlebel)*prior_pos
likeprior_pos=multivariate_normal_pos.pdf(demo_covid_train_withoutlebel)*prior_pos
likeprior_neg=multivariate_normal_neg.pdf(demo_covid_train_withoutlebel)*prior_neg
post_pos=likeprior_pos/likeprior_pos+likeprior_neg
post_neg=likeprior_neg/likeprior_pos+likeprior_neg

#If below condition satisfied then it will be true otherwise false. This nway control thresholed kept as 0.5.

In [28]:
pred= post_pos>post_neg
pred

array([False, False,  True, False,  True, False, False,  True,  True,
        True, False,  True,  True,  True, False,  True, False, False,
        True, False, False, False,  True, False,  True, False,  True,
       False,  True, False, False, False, False, False, False, False,
        True, False, False,  True,  True,  True, False,  True,  True,
        True, False, False, False, False, False, False, False,  True,
        True, False,  True, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False,  True, False, False,
       False, False, False,  True, False,  True, False,  True, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False, False,
       False, False, False,  True, False, False,  True, False, False,
       False, False, False,  True, False, False, False, False,  True,
       False, False,

In [31]:
labels = demo_covid_train['9']
labels

0       True
1      False
2       True
3      False
4       True
5      False
6       True
7      False
8       True
9       True
10     False
11      True
12     False
13      True
14      True
15      True
16      True
17      True
18     False
19      True
20     False
21     False
22      True
23      True
24      True
25      True
26      True
27     False
28     False
29     False
       ...  
470    False
471    False
472    False
473    False
474    False
475    False
476     True
477    False
478    False
479    False
480     True
481    False
482    False
483    False
484     True
485     True
486    False
487    False
488    False
489    False
490    False
491    False
492    False
493     True
494    False
495    False
496    False
497    False
498     True
499    False
Name: 9, Length: 500, dtype: object

In [24]:
labels[labels=='pos'] = True
labels[labels=='neg'] =False

C:\Users\Dell-672206\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\Dell-672206\Anaconda3\lib\site-packages\pandas\core\generic.py:8682: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
C:\Users\Dell-672206\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_n

In [25]:
labels

0       True
1      False
2       True
3      False
4       True
5      False
6       True
7      False
8       True
9       True
10     False
11      True
12     False
13      True
14      True
15      True
16      True
17      True
18     False
19      True
20     False
21     False
22      True
23      True
24      True
25      True
26      True
27     False
28     False
29     False
       ...  
470    False
471    False
472    False
473    False
474    False
475    False
476     True
477    False
478    False
479    False
480     True
481    False
482    False
483    False
484     True
485     True
486    False
487    False
488    False
489    False
490    False
491    False
492    False
493     True
494    False
495    False
496    False
497    False
498     True
499    False
Name: 9, Length: 500, dtype: object

Training Error

if lebels & Predict have same string then = sum their numbers. 498 index of both pred &  lebels is True.But 499 pred is true Lebel is false. Then divide by total number of that string in Pred. This will give accuracy.


In [27]:
(sum((labels == pred))/len(pred))*100

75.4

Testing

In [30]:
demo_covid_test_withoutlebel= demo_covid_test.drop(['9'],axis=1)
likeprior_pos=multivariate_normal_pos.pdf(demo_covid_test_withoutlebel)*prior_pos
likeprior_neg=multivariate_normal_neg.pdf(demo_covid_test_withoutlebel)*prior_neg
post_pos=likeprior_pos/likeprior_pos+likeprior_neg
post_neg=likeprior_neg/likeprior_pos+likeprior_neg

In [32]:
pred= post_pos>post_neg

In [33]:
labels = demo_covid_test['9']
labels[labels=='pos'] = True
labels[labels=='neg'] = False

C:\Users\Dell-672206\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\Dell-672206\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [34]:
(sum((labels == pred))/len(pred))*100

77.98507462686567

Testing you are COVID+ve or Negative

In [53]:
% patient data
%            [pregnencies, glucose,blood pressure, skin thickness, insulin, BMI, Diabetese Pedigree function, age] 
patient_data=[0,170,80,20,20,20,0.5,50]


UsageError: Line magic function `%` not found.


In [54]:
likeprior_pos=multivariate_normal_pos.pdf(patient_data)*prior_pos
likeprior_neg=multivariate_normal_neg.pdf(patient_data)*prior_neg
post_pos=likeprior_pos/likeprior_pos+likeprior_neg
post_neg=likeprior_neg/likeprior_pos+likeprior_neg

In [55]:
pred= post_pos>post_neg


In [56]:
pred

False